File: 03b.py
#!/usr/bin/python3
import sys

priority = dict([(chr(c), 1 + c - ord('a'))
                 for c in range(ord('a'), 1 + ord('z'))]
                + [(chr(c), 27 + c - ord('A'))
                   for c in range(ord('A'), 1 + ord('Z'))])

total = 0
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        line = line.rstrip('\n')
        common = set(line)
        for i in (1, 2):
            following = next(fh, None)
            common = common.intersection(set(following))
        total += priority[list(common)[0]]
print(total)

File: 03b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my %PRIORITY = map { $_    => ord($_) - ord '`',
                     uc $_ =>  ord($_) - 70
               } 'a' .. 'z';

my $sum;
while (1) {
    my @group = map scalar <>, 1 .. 3;
    chomp @group;
    my %score;
    for my $g (0 .. 2) {
        undef $score{$_}{$g} for split //, $group[$g];
    }
    my $common = (grep 3 == keys %{ $score{$_} }, keys %score)[0];
    $sum += $PRIORITY{$common};
    last if eof;
}
say $sum;

__DATA__
vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
